src/parser/datetime.rs

```diff
@@ -104,14 +194,115 @@ fn potential_interval_tokens(from: &DateTimeField) -> Vec<IntervalToken> {
+        vec![Plus, Num(0), Num(0)],
+        vec![Dash, Num(0), Num(0)],
+        vec![Plus, Num(0)],
+        vec![Dash, Num(0)],
+        vec![TzName("".to_string())],
+        vec![Zulu],
+    ];
+
+    let mut is_positive = true;
+    let mut hour_offset: Option<i64> = None;
+    let mut minute_offset: Option<i64> = None;
+
+    for format in all_formats.iter() {
+        let actual = tokens.iter();
+
+        if actual.len() != format.len() {
+            continue;
+        }
+
+        for (i, (atok, etok)) in actual.zip(format).enumerate() {
+            match (atok, etok) {
+                (Colon, Colon) | (Plus, Plus) => { /* Matching punctuation */ }
+                (Dash, Dash) => {
+                    is_positive = false;
+                }
+                (Num(val), Num(_)) => {
+                    let val = *val;
+                    if hour_offset.is_none() {
+                        // We haven't seen an hour offset yet so lets try to parse that
+                        if val <= 24 {
+                            hour_offset = Some(val as i64);
+                        } else {
+                            return Err(ParserError::ParserError(format!(
+                                "Error parsing timezone string ({}): timezone hour invalid {}",
+                                value, val
+                            )));
+                        }
+                    } else if minute_offset.is_none() {
+                        if val <= 60 {
+                            minute_offset = Some(val as i64);
+                        } else {
+                            return Err(ParserError::ParserError(format!(
+                                "Error parsing timezone string ({}): timezone minute invalid {}",
+                                value, val
+                            )));
+                        }
+                    } else {
+                        // We've already seen an hour and a minute so we should never see another number
+                        return Err(ParserError::ParserError(format!(
+                            "Error parsing timezone string ({}): invalid value {} at token index {}", value,
+                            val, i
+                        )));
+                    }
+                }
+                (Zulu, Zulu) => return Ok(0 as i64),
+                (TzName(val), TzName(_)) => {
+                    // For now, we don't support named timezones
+                    return Err(ParserError::ParserError(format!(
+                        "Error parsing timezone string ({}): named timezones are not supported. \
+                         Failed to parse {} at token index {}",
+                        value, val, i
+                    )));
+                }
+                (_, _) => {
+                    is_positive = true;
+                    hour_offset = None;
+                    minute_offset = None;
+                    break;
+                }
```

Comment:
Even though I recognize that this is obviously working, I don't think that I follow it. Does it unconditionally go through every format string, even if the first one matches? That would explain why this needs to be at the end instead of the beginning. Would it be easier to reason about this if this inner match statement was in an inner function that can signal success by returning a success/fail value?

The reason this feels worth doing is that I _believe_ that the entire function currently depends on the order of items in that format list plus the order of items in this match, and is therefore possibly pretty fragile in the face of refactorings. The tests will prevent people from introducing incorrect code, but I could see trying to interact with this code requiring a lot of understanding.

If you don't feel like experimenting with a refactoring, could you add a comment at the top of the possible formats loop (line 217) explaining the ways that these parts all fit together?

Reply:
So the intention was to match on the first format that validly parses (the structure is a doubly nested for loop so the `break` statement is there when we've detected a mismatch between the format and the actual list of tokens and lets the control flow go to the next one), with the understanding that all of the potential formats with numbers are going to have numbers in HH MM order, so if the first number is >= 24, then we can just fail everything because we know it will never pass any format (as opposed to a token mismatch which could pass a different format)

I'll refactor a bit and add comments

Reply:
thanks!